java.lang.AssertionError: assertion failed: Attempt to commit TS[11888X, 11724X, 11481X, 11061X, 10149, 2,

## Compiler version

Tried both `3.0.1` and `3.0.2-RC1`

## Minimized code

See below
https://github.com/lampepfl/dotty/issues/13101#issuecomment-882932439

## Output (click arrow to expand)
<details>

`sbt compile`

```shell
[info] exception occurred while compiling /Users/swoorup.joshi/personal/zicoin/src/main/scala/Main.scala, /Users/swoorup.joshi/personal/zicoin/src/main/scala/actor/Blockchain.scala, /Users/swoorup.joshi/personal/zicoin/src/main/scala/actor/Broker.scala, /Users/swoorup.joshi/personal/zicoin/src/main/scala/actor/Miner.scala, /Users/swoorup.joshi/personal/zicoin/src/main/scala/actor/Node.scala, /Users/swoorup.joshi/personal/zicoin/src/main/scala/api/NodeRoutes.scala, /Users/swoorup.joshi/personal/zicoin/src/main/scala/blockchain/Chain.scala, /Users/swoorup.joshi/personal/zicoin/src/main/scala/blockchain/Core.scala, /Users/swoorup.joshi/personal/zicoin/src/main/scala/blockchain/Transaction.scala, /Users/swoorup.joshi/personal/zicoin/src/main/scala/cluster/ClusterListener.scala, /Users/swoorup.joshi/personal/zicoin/src/main/scala/cluster/ClusterManager.scala, /Users/swoorup.joshi/personal/zicoin/src/main/scala/common/Common.scala, /Users/swoorup.joshi/personal/zicoin/src/main/scala/common/NewType.scala, /Users/swoorup.joshi/personal/zicoin/src/main/scala/crypto/Crypto.scala, /Users/swoorup.joshi/personal/zicoin/src/main/scala/de/heikoseeberger/akkahttpcirce/CirceSupport.scala, /Users/swoorup.joshi/personal/zicoin/src/main/scala/exception/InvalidProofException.scala, /Users/swoorup.joshi/personal/zicoin/src/main/scala/exception/MinerBusyException.scala, /Users/swoorup.joshi/personal/zicoin/src/main/scala/proof/ProofOfWork.scala
java.lang.AssertionError: assertion failed: Attempt to commit TS[11888X, 11724X, 11481X, 11061X, 10149, 2, 1, 0] into already committed TS[11724X, 11481X, 11061X, 10149, 2, 1, 0] while compiling /Users/swoorup.joshi/personal/zicoin/src/main/scala/Main.scala, /Users/swoorup.joshi/personal/zicoin/src/main/scala/actor/Blockchain.scala, /Users/swoorup.joshi/personal/zicoin/src/main/scala/actor/Broker.scala, /Users/swoorup.joshi/personal/zicoin/src/main/scala/actor/Miner.scala, /Users/swoorup.joshi/personal/zicoin/src/main/scala/actor/Node.scala, /Users/swoorup.joshi/personal/zicoin/src/main/scala/api/NodeRoutes.scala, /Users/swoorup.joshi/personal/zicoin/src/main/scala/blockchain/Chain.scala, /Users/swoorup.joshi/personal/zicoin/src/main/scala/blockchain/Core.scala, /Users/swoorup.joshi/personal/zicoin/src/main/scala/blockchain/Transaction.scala, /Users/swoorup.joshi/personal/zicoin/src/main/scala/cluster/ClusterListener.scala, /Users/swoorup.joshi/personal/zicoin/src/main/scala/cluster/ClusterManager.scala, /Users/swoorup.joshi/personal/zicoin/src/main/scala/common/Common.scala, /Users/swoorup.joshi/personal/zicoin/src/main/scala/common/NewType.scala, /Users/swoorup.joshi/personal/zicoin/src/main/scala/crypto/Crypto.scala, /Users/swoorup.joshi/personal/zicoin/src/main/scala/de/heikoseeberger/akkahttpcirce/CirceSupport.scala, /Users/swoorup.joshi/personal/zicoin/src/main/scala/exception/InvalidProofException.scala, /Users/swoorup.joshi/personal/zicoin/src/main/scala/exception/MinerBusyException.scala, /Users/swoorup.joshi/personal/zicoin/src/main/scala/proof/ProofOfWork.scala
[error] ## Exception when compiling 18 sources to /Users/swoorup.joshi/personal/zicoin/target/scala-3.0.2-RC1/classes
[error] java.lang.AssertionError: assertion failed: Attempt to commit TS[11888X, 11724X, 11481X, 11061X, 10149, 2, 1, 0] into already committed TS[11724X, 11481X, 11061X, 10149, 2, 1, 0]
[error] scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:8)
[error] dotty.tools.dotc.core.TyperState.commit(TyperState.scala:146)
[error] dotty.tools.dotc.typer.Inferencing$.isFullyDefined(Inferencing.scala:40)
[error] dotty.tools.dotc.typer.Implicits.ignoredInstanceNormalImport$1(Implicits.scala:881)
[error] dotty.tools.dotc.typer.Implicits.$anonfun$1(Implicits.scala:896)
[error] dotty.tools.dotc.typer.ImplicitSearchError.hiddenImplicitsAddendum(ErrorReporting.scala:419)
[error] dotty.tools.dotc.typer.ImplicitSearchError.missingArgMsg(ErrorReporting.scala:267)
[error] dotty.tools.dotc.typer.Implicits.missingArgMsg(Implicits.scala:897)
[error] dotty.tools.dotc.typer.Implicits.missingArgMsg$(Implicits.scala:774)
[error] dotty.tools.dotc.typer.Typer.missingArgMsg(Typer.scala:106)
[error] dotty.tools.dotc.typer.Typer.issueErrors$1$$anonfun$1$$anonfun$1(Typer.scala:3353)
[error] dotty.tools.dotc.reporting.NoExplanation.msg(Message.scala:132)
[error] dotty.tools.dotc.reporting.Message.message(Message.scala:87)
[error] dotty.tools.dotc.reporting.Message.isNonSensical(Message.scala:99)
[error] dotty.tools.dotc.reporting.HideNonSensicalMessages.isHidden(HideNonSensicalMessages.scala:16)
[error] dotty.tools.dotc.reporting.HideNonSensicalMessages.isHidden$(HideNonSensicalMessages.scala:10)
[error] dotty.tools.dotc.reporting.AbstractReporter.isHidden(AbstractReporter.scala:8)
[error] dotty.tools.dotc.reporting.Reporter.report(Reporter.scala:148)
[error] dotty.tools.dotc.reporting.Reporter.flush$$anonfun$1(Reporter.scala:206)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.immutable.List.foreach(List.scala:333)
[error] dotty.tools.dotc.reporting.Reporter.flush(Reporter.scala:206)
[error] dotty.tools.dotc.core.TyperState.commit(TyperState.scala:143)
[error] dotty.tools.dotc.typer.Applications.fail$1(Applications.scala:986)
[error] dotty.tools.dotc.typer.Applications.realApply$7$$anonfun$6$$anonfun$2(Applications.scala:992)
[error] scala.Option.getOrElse(Option.scala:201)
[error] dotty.tools.dotc.typer.Applications.realApply$8$$anonfun$7(Applications.scala:993)
[error] dotty.tools.dotc.typer.Typer.tryEither(Typer.scala:2952)
[error] dotty.tools.dotc.typer.Applications.realApply$1(Applications.scala:994)
[error] dotty.tools.dotc.typer.Applications.typedApply(Applications.scala:1032)
[error] dotty.tools.dotc.typer.Applications.typedApply$(Applications.scala:317)
[error] dotty.tools.dotc.typer.Typer.typedApply(Typer.scala:106)
[error] dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2693)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2756)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2821)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2825)
[error] dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:2941)
[error] dotty.tools.dotc.typer.Typer.typedBlock(Typer.scala:1031)
[error] dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2701)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2756)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2821)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2825)
[error] dotty.tools.dotc.typer.ProtoTypes$FunProto.$anonfun$6$$anonfun$1(ProtoTypes.scala:384)
[error] dotty.tools.dotc.typer.ProtoTypes$FunProto.cacheTypedArg(ProtoTypes.scala:359)
[error] dotty.tools.dotc.typer.ProtoTypes$FunProto.$anonfun$7(ProtoTypes.scala:384)
[error] dotty.tools.dotc.typer.ProtoTypes$FunProto.$anonfun$adapted$1(ProtoTypes.scala:384)
[error] dotty.tools.dotc.core.Decorators$ListDecorator$.loop$3(Decorators.scala:179)
[error] dotty.tools.dotc.core.Decorators$ListDecorator$.mapWithIndexConserve$extension(Decorators.scala:186)
[error] dotty.tools.dotc.typer.ProtoTypes$FunProto.typedArgs(ProtoTypes.scala:384)
[error] dotty.tools.dotc.typer.ProtoTypes$FunProto.isMatchedBy(ProtoTypes.scala:314)
[error] dotty.tools.dotc.core.TypeComparer.isMatchedByProto(TypeComparer.scala:1874)
[error] dotty.tools.dotc.core.TypeComparer.firstTry$1(TypeComparer.scala:315)
[error] dotty.tools.dotc.core.TypeComparer.recur(TypeComparer.scala:1296)
[error] dotty.tools.dotc.core.TypeComparer.isSubType(TypeComparer.scala:195)
[error] dotty.tools.dotc.core.TypeComparer.isSubType(TypeComparer.scala:205)
[error] dotty.tools.dotc.core.TypeComparer.topLevelSubType(TypeComparer.scala:128)
[error] dotty.tools.dotc.core.TypeComparer$.topLevelSubType(TypeComparer.scala:2700)
[error] dotty.tools.dotc.core.Types$Type.$less$colon$less(Types.scala:1032)
[error] dotty.tools.dotc.core.Types$Type.relaxed_$less$colon$less(Types.scala:1064)
[error] dotty.tools.dotc.typer.ProtoTypes$Compatibility.isCompatible(ProtoTypes.scala:39)
[error] dotty.tools.dotc.typer.ProtoTypes$Compatibility.isCompatible$(ProtoTypes.scala:26)
[error] dotty.tools.dotc.typer.ProtoTypes$NoViewsAllowed$.isCompatible(ProtoTypes.scala:120)
[error] dotty.tools.dotc.typer.ProtoTypes$Compatibility.testCompat$1(ProtoTypes.scala:56)
[error] dotty.tools.dotc.typer.ProtoTypes$Compatibility.normalizedCompatible(ProtoTypes.scala:71)
[error] dotty.tools.dotc.typer.ProtoTypes$Compatibility.normalizedCompatible$(ProtoTypes.scala:26)
[error] dotty.tools.dotc.typer.ProtoTypes$NoViewsAllowed$.normalizedCompatible(ProtoTypes.scala:120)
[error] dotty.tools.dotc.typer.ProtoTypes$SelectionProto.qualifies$1(ProtoTypes.scala:192)
[error] dotty.tools.dotc.typer.ProtoTypes$SelectionProto.isMatchedBy(ProtoTypes.scala:196)
[error] dotty.tools.dotc.typer.Typer.tryInsertImplicitOnQualifier(Typer.scala:3050)
[error] dotty.tools.dotc.typer.Applications.tryWithImplicitOnQualifier$4(Applications.scala:921)
[error] dotty.tools.dotc.typer.Applications.realApply$8$$anonfun$7(Applications.scala:991)
[error] dotty.tools.dotc.typer.Typer.tryEither(Typer.scala:2952)
[error] dotty.tools.dotc.typer.Applications.realApply$1(Applications.scala:994)
[error] dotty.tools.dotc.typer.Applications.typedApply(Applications.scala:1032)
[error] dotty.tools.dotc.typer.Applications.typedApply$(Applications.scala:317)
[error] dotty.tools.dotc.typer.Typer.typedApply(Typer.scala:106)
[error] dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2693)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2756)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2821)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2825)
[error] dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:2941)
[error] dotty.tools.dotc.typer.Typer.typedValDef(Typer.scala:2092)
[error] dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2667)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2755)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2821)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2825)
[error] dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:2847)
[error] dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:2897)
[error] dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:2352)
[error] dotty.tools.dotc.typer.Typer.typedTypeOrClassDef$2(Typer.scala:2681)
[error] dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2685)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2755)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2821)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2825)
[error] dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:2847)
[error] dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:2897)
[error] dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:2475)
[error] dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2726)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2756)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2821)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2825)
[error] dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:2941)
[error] dotty.tools.dotc.typer.FrontEnd.liftedTree1$1(FrontEnd.scala:79)
[error] dotty.tools.dotc.typer.FrontEnd.typeCheck$$anonfun$1(FrontEnd.scala:84)
[error] dotty.tools.dotc.typer.FrontEnd.monitor(FrontEnd.scala:43)
[error] dotty.tools.dotc.typer.FrontEnd.typeCheck(FrontEnd.scala:85)
[error] dotty.tools.dotc.typer.FrontEnd.runOn$$anonfun$3(FrontEnd.scala:120)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.immutable.List.foreach(List.scala:333)
[error] dotty.tools.dotc.typer.FrontEnd.runOn(FrontEnd.scala:120)
[error] dotty.tools.dotc.Run.runPhases$4$$anonfun$4(Run.scala:205)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
[error] dotty.tools.dotc.Run.runPhases$5(Run.scala:216)
[error] dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:224)
[error] scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
[error] dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:67)
[error] dotty.tools.dotc.Run.compileUnits(Run.scala:231)
[error] dotty.tools.dotc.Run.compileSources(Run.scala:166)
[error] dotty.tools.dotc.Run.compile(Run.scala:150)
[error] dotty.tools.dotc.Driver.doCompile(Driver.scala:39)
[error] dotty.tools.xsbt.CompilerBridgeDriver.run(CompilerBridgeDriver.java:88)
[error] dotty.tools.xsbt.CompilerBridge.run(CompilerBridge.java:22)
[error] sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:91)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$7(MixedAnalyzingCompiler.scala:192)
[error] scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[error] sbt.internal.inc.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:247)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4(MixedAnalyzingCompiler.scala:182)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4$adapted(MixedAnalyzingCompiler.scala:163)
[error] sbt.internal.inc.JarUtils$.withPreviousJar(JarUtils.scala:239)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compileScala$1(MixedAnalyzingCompiler.scala:163)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:210)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1(IncrementalCompilerImpl.scala:528)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1$adapted(IncrementalCompilerImpl.scala:528)
[error] sbt.internal.inc.Incremental$.$anonfun$apply$5(Incremental.scala:175)
[error] sbt.internal.inc.Incremental$.$anonfun$apply$5$adapted(Incremental.scala:173)
[error] sbt.internal.inc.Incremental$$anon$2.run(Incremental.scala:459)
[error] sbt.internal.inc.IncrementalCommon$CycleState.next(IncrementalCommon.scala:116)
[error] sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:56)
[error] sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:52)
[error] sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:263)
[error] sbt.internal.inc.Incremental$.$anonfun$incrementalCompile$8(Incremental.scala:414)
[error] sbt.internal.inc.Incremental$.withClassfileManager(Incremental.scala:501)
[error] sbt.internal.inc.Incremental$.incrementalCompile(Incremental.scala:401)
[error] sbt.internal.inc.Incremental$.apply(Incremental.scala:167)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:528)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1(IncrementalCompilerImpl.scala:482)
[error] sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError(IncrementalCompilerImpl.scala:332)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally(IncrementalCompilerImpl.scala:420)
[error] sbt.internal.inc.IncrementalCompilerImpl.compile(IncrementalCompilerImpl.scala:137)
[error] sbt.Defaults$.compileIncrementalTaskImpl(Defaults.scala:2357)
[error] sbt.Defaults$.$anonfun$compileIncrementalTask$2(Defaults.scala:2314)
[error] sbt.internal.io.Retry$.apply(Retry.scala:46)
[error] sbt.internal.io.Retry$.apply(Retry.scala:28)
[error] sbt.internal.io.Retry$.apply(Retry.scala:23)
[error] sbt.internal.server.BspCompileTask$.compute(BspCompileTask.scala:31)
[error] sbt.Defaults$.$anonfun$compileIncrementalTask$1(Defaults.scala:2310)
[error] scala.Function1.$anonfun$compose$1(Function1.scala:49)
[error] sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:62)
[error] sbt.std.Transform$$anon$4.work(Transform.scala:68)
[error] sbt.Execute.$anonfun$submit$2(Execute.scala:282)
[error] sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:23)
[error] sbt.Execute.work(Execute.scala:291)
[error] sbt.Execute.$anonfun$submit$1(Execute.scala:282)
[error] sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:265)
[error] sbt.CompletionService$$anon$2.call(CompletionService.scala:64)
[error] java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[error] java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
[error] java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[error] java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
[error] java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
[error] java.base/java.lang.Thread.run(Thread.java:834)
[error]
[error] java.lang.AssertionError: assertion failed: Attempt to commit TS[11888X, 11724X, 11481X, 11061X, 10149, 2, 1, 0] into already committed TS[11724X, 11481X, 11061X, 10149, 2, 1, 0]
[error]         at scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:8)
[error]         at dotty.tools.dotc.core.TyperState.commit(TyperState.scala:146)
[error]         at dotty.tools.dotc.typer.Inferencing$.isFullyDefined(Inferencing.scala:40)
[error]         at dotty.tools.dotc.typer.Implicits.ignoredInstanceNormalImport$1(Implicits.scala:881)
[error]         at dotty.tools.dotc.typer.Implicits.$anonfun$1(Implicits.scala:896)
[error]         at dotty.tools.dotc.typer.ImplicitSearchError.hiddenImplicitsAddendum(ErrorReporting.scala:419)
[error]         at dotty.tools.dotc.typer.ImplicitSearchError.missingArgMsg(ErrorReporting.scala:267)
[error]         at dotty.tools.dotc.typer.Implicits.missingArgMsg(Implicits.scala:897)
[error]         at dotty.tools.dotc.typer.Implicits.missingArgMsg$(Implicits.scala:774)
[error]         at dotty.tools.dotc.typer.Typer.missingArgMsg(Typer.scala:106)
[error]         at dotty.tools.dotc.typer.Typer.issueErrors$1$$anonfun$1$$anonfun$1(Typer.scala:3353)
[error]         at dotty.tools.dotc.reporting.NoExplanation.msg(Message.scala:132)
[error]         at dotty.tools.dotc.reporting.Message.message(Message.scala:87)
[error]         at dotty.tools.dotc.reporting.Message.isNonSensical(Message.scala:99)
[error]         at dotty.tools.dotc.reporting.HideNonSensicalMessages.isHidden(HideNonSensicalMessages.scala:16)
[error]         at dotty.tools.dotc.reporting.HideNonSensicalMessages.isHidden$(HideNonSensicalMessages.scala:10)
[error]         at dotty.tools.dotc.reporting.AbstractReporter.isHidden(AbstractReporter.scala:8)
[error]         at dotty.tools.dotc.reporting.Reporter.report(Reporter.scala:148)
[error]         at dotty.tools.dotc.reporting.Reporter.flush$$anonfun$1(Reporter.scala:206)
[error]         at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error]         at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error]         at scala.collection.immutable.List.foreach(List.scala:333)
[error]         at dotty.tools.dotc.reporting.Reporter.flush(Reporter.scala:206)
[error]         at dotty.tools.dotc.core.TyperState.commit(TyperState.scala:143)
[error]         at dotty.tools.dotc.typer.Applications.fail$1(Applications.scala:986)
[error]         at dotty.tools.dotc.typer.Applications.realApply$7$$anonfun$6$$anonfun$2(Applications.scala:992)
[error]         at scala.Option.getOrElse(Option.scala:201)
[error]         at dotty.tools.dotc.typer.Applications.realApply$8$$anonfun$7(Applications.scala:993)
[error]         at dotty.tools.dotc.typer.Typer.tryEither(Typer.scala:2952)
[error]         at dotty.tools.dotc.typer.Applications.realApply$1(Applications.scala:994)
[error]         at dotty.tools.dotc.typer.Applications.typedApply(Applications.scala:1032)
[error]         at dotty.tools.dotc.typer.Applications.typedApply$(Applications.scala:317)
[error]         at dotty.tools.dotc.typer.Typer.typedApply(Typer.scala:106)
[error]         at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2693)
[error]         at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2756)
[error]         at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2821)
[error]         at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2825)
[error]         at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:2941)
[error]         at dotty.tools.dotc.typer.Typer.typedBlock(Typer.scala:1031)
[error]         at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2701)
[error]         at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2756)
[error]         at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2821)
[error]         at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2825)
[error]         at dotty.tools.dotc.typer.ProtoTypes$FunProto.$anonfun$6$$anonfun$1(ProtoTypes.scala:384)
[error]         at dotty.tools.dotc.typer.ProtoTypes$FunProto.cacheTypedArg(ProtoTypes.scala:359)
[error]         at dotty.tools.dotc.typer.ProtoTypes$FunProto.$anonfun$7(ProtoTypes.scala:384)
[error]         at dotty.tools.dotc.typer.ProtoTypes$FunProto.$anonfun$adapted$1(ProtoTypes.scala:384)
[error]         at dotty.tools.dotc.core.Decorators$ListDecorator$.loop$3(Decorators.scala:179)
[error]         at dotty.tools.dotc.core.Decorators$ListDecorator$.mapWithIndexConserve$extension(Decorators.scala:186)
[error]         at dotty.tools.dotc.typer.ProtoTypes$FunProto.typedArgs(ProtoTypes.scala:384)
[error]         at dotty.tools.dotc.typer.ProtoTypes$FunProto.isMatchedBy(ProtoTypes.scala:314)
[error]         at dotty.tools.dotc.core.TypeComparer.isMatchedByProto(TypeComparer.scala:1874)
[error]         at dotty.tools.dotc.core.TypeComparer.firstTry$1(TypeComparer.scala:315)
[error]         at dotty.tools.dotc.core.TypeComparer.recur(TypeComparer.scala:1296)
[error]         at dotty.tools.dotc.core.TypeComparer.isSubType(TypeComparer.scala:195)
[error]         at dotty.tools.dotc.core.TypeComparer.isSubType(TypeComparer.scala:205)
[error]         at dotty.tools.dotc.core.TypeComparer.topLevelSubType(TypeComparer.scala:128)
[error]         at dotty.tools.dotc.core.TypeComparer$.topLevelSubType(TypeComparer.scala:2700)
[error]         at dotty.tools.dotc.core.Types$Type.$less$colon$less(Types.scala:1032)
[error]         at dotty.tools.dotc.core.Types$Type.relaxed_$less$colon$less(Types.scala:1064)
[error]         at dotty.tools.dotc.typer.ProtoTypes$Compatibility.isCompatible(ProtoTypes.scala:39)
[error]         at dotty.tools.dotc.typer.ProtoTypes$Compatibility.isCompatible$(ProtoTypes.scala:26)
[error]         at dotty.tools.dotc.typer.ProtoTypes$NoViewsAllowed$.isCompatible(ProtoTypes.scala:120)
[error]         at dotty.tools.dotc.typer.ProtoTypes$Compatibility.testCompat$1(ProtoTypes.scala:56)
[error]         at dotty.tools.dotc.typer.ProtoTypes$Compatibility.normalizedCompatible(ProtoTypes.scala:71)
[error]         at dotty.tools.dotc.typer.ProtoTypes$Compatibility.normalizedCompatible$(ProtoTypes.scala:26)
[error]         at dotty.tools.dotc.typer.ProtoTypes$NoViewsAllowed$.normalizedCompatible(ProtoTypes.scala:120)
[error]         at dotty.tools.dotc.typer.ProtoTypes$SelectionProto.qualifies$1(ProtoTypes.scala:192)
[error]         at dotty.tools.dotc.typer.ProtoTypes$SelectionProto.isMatchedBy(ProtoTypes.scala:196)
[error]         at dotty.tools.dotc.typer.Typer.tryInsertImplicitOnQualifier(Typer.scala:3050)
[error]         at dotty.tools.dotc.typer.Applications.tryWithImplicitOnQualifier$4(Applications.scala:921)
[error]         at dotty.tools.dotc.typer.Applications.realApply$8$$anonfun$7(Applications.scala:991)
[error]         at dotty.tools.dotc.typer.Typer.tryEither(Typer.scala:2952)
[error]         at dotty.tools.dotc.typer.Applications.realApply$1(Applications.scala:994)
[error]         at dotty.tools.dotc.typer.Applications.typedApply(Applications.scala:1032)
[error]         at dotty.tools.dotc.typer.Applications.typedApply$(Applications.scala:317)
[error]         at dotty.tools.dotc.typer.Typer.typedApply(Typer.scala:106)
[error]         at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2693)
[error]         at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2756)
[error]         at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2821)
[error]         at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2825)
[error]         at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:2941)
[error]         at dotty.tools.dotc.typer.Typer.typedValDef(Typer.scala:2092)
[error]         at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2667)
[error]         at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2755)
[error]         at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2821)
[error]         at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2825)
[error]         at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:2847)
[error]         at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:2897)
[error]         at dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:2352)
[error]         at dotty.tools.dotc.typer.Typer.typedTypeOrClassDef$2(Typer.scala:2681)
[error]         at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2685)
[error]         at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2755)
[error]         at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2821)
[error]         at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2825)
[error]         at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:2847)
[error]         at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:2897)
[error]         at dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:2475)
[error]         at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2726)
[error]         at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2756)
[error]         at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2821)
[error]         at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2825)
[error]         at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:2941)
[error]         at dotty.tools.dotc.typer.FrontEnd.liftedTree1$1(FrontEnd.scala:79)
[error]         at dotty.tools.dotc.typer.FrontEnd.typeCheck$$anonfun$1(FrontEnd.scala:84)
[error]         at dotty.tools.dotc.typer.FrontEnd.monitor(FrontEnd.scala:43)
[error]         at dotty.tools.dotc.typer.FrontEnd.typeCheck(FrontEnd.scala:85)
[error]         at dotty.tools.dotc.typer.FrontEnd.runOn$$anonfun$3(FrontEnd.scala:120)
[error]         at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error]         at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error]         at scala.collection.immutable.List.foreach(List.scala:333)
[error]         at dotty.tools.dotc.typer.FrontEnd.runOn(FrontEnd.scala:120)
[error]         at dotty.tools.dotc.Run.runPhases$4$$anonfun$4(Run.scala:205)
[error]         at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error]         at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error]         at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
[error]         at dotty.tools.dotc.Run.runPhases$5(Run.scala:216)
[error]         at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:224)
[error]         at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
[error]         at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:67)
[error]         at dotty.tools.dotc.Run.compileUnits(Run.scala:231)
[error]         at dotty.tools.dotc.Run.compileSources(Run.scala:166)
[error]         at dotty.tools.dotc.Run.compile(Run.scala:150)
[error]         at dotty.tools.dotc.Driver.doCompile(Driver.scala:39)
[error]         at dotty.tools.xsbt.CompilerBridgeDriver.run(CompilerBridgeDriver.java:88)
[error]         at dotty.tools.xsbt.CompilerBridge.run(CompilerBridge.java:22)
[error]         at sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:91)
[error]         at sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$7(MixedAnalyzingCompiler.scala:192)
[error]         at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[error]         at sbt.internal.inc.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:247)
[error]         at sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4(MixedAnalyzingCompiler.scala:182)
[error]         at sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4$adapted(MixedAnalyzingCompiler.scala:163)
[error]         at sbt.internal.inc.JarUtils$.withPreviousJar(JarUtils.scala:239)
[error]         at sbt.internal.inc.MixedAnalyzingCompiler.compileScala$1(MixedAnalyzingCompiler.scala:163)
[error]         at sbt.internal.inc.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:210)
[error]         at sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1(IncrementalCompilerImpl.scala:528)
[error]         at sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1$adapted(IncrementalCompilerImpl.scala:528)
[error]         at sbt.internal.inc.Incremental$.$anonfun$apply$5(Incremental.scala:175)
[error]         at sbt.internal.inc.Incremental$.$anonfun$apply$5$adapted(Incremental.scala:173)
[error]         at sbt.internal.inc.Incremental$$anon$2.run(Incremental.scala:459)
[error]         at sbt.internal.inc.IncrementalCommon$CycleState.next(IncrementalCommon.scala:116)
[error]         at sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:56)
[error]         at sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:52)
[error]         at sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:263)
[error]         at sbt.internal.inc.Incremental$.$anonfun$incrementalCompile$8(Incremental.scala:414)
[error]         at sbt.internal.inc.Incremental$.withClassfileManager(Incremental.scala:501)
[error]         at sbt.internal.inc.Incremental$.incrementalCompile(Incremental.scala:401)
[error]         at sbt.internal.inc.Incremental$.apply(Incremental.scala:167)
[error]         at sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:528)
[error]         at sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1(IncrementalCompilerImpl.scala:482)
[error]         at sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError(IncrementalCompilerImpl.scala:332)
[error]         at sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally(IncrementalCompilerImpl.scala:420)
[error]         at sbt.internal.inc.IncrementalCompilerImpl.compile(IncrementalCompilerImpl.scala:137)
[error]         at sbt.Defaults$.compileIncrementalTaskImpl(Defaults.scala:2357)
[error]         at sbt.Defaults$.$anonfun$compileIncrementalTask$2(Defaults.scala:2314)
[error]         at sbt.internal.io.Retry$.apply(Retry.scala:46)
[error]         at sbt.internal.io.Retry$.apply(Retry.scala:28)
[error]         at sbt.internal.io.Retry$.apply(Retry.scala:23)
[error]         at sbt.internal.server.BspCompileTask$.compute(BspCompileTask.scala:31)
[error]         at sbt.Defaults$.$anonfun$compileIncrementalTask$1(Defaults.scala:2310)
[error]         at scala.Function1.$anonfun$compose$1(Function1.scala:49)
[error]         at sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:62)
[error]         at sbt.std.Transform$$anon$4.work(Transform.scala:68)
[error]         at sbt.Execute.$anonfun$submit$2(Execute.scala:282)
[error]         at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:23)
[error]         at sbt.Execute.work(Execute.scala:291)
[error]         at sbt.Execute.$anonfun$submit$1(Execute.scala:282)
[error]         at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:265)
[error]         at sbt.CompletionService$$anon$2.call(CompletionService.scala:64)
[error]         at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[error]         at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
[error]         at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[error]         at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
[error]         at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
[error]         at java.base/java.lang.Thread.run(Thread.java:834)
[error] (Compile / compileIncremental) java.lang.AssertionError: assertion failed: Attempt to commit TS[11888X, 11724X, 11481X, 11061X, 10149, 2, 1, 0] into already committed TS[11724X, 11481X, 11061X, 10149, 2, 1, 0]
[error] Total time: 6 s, completed 19 Jul. 2021, 7:39:34 pm

```
</details>

A self-contained reproducer:

```scala
trait Vehicle
trait Car extends Vehicle

trait Encoder[A]
object Encoder {
  implicit val encodeVehicle: Encoder[Vehicle] = ???
  implicit val encodeCar: Encoder[Car] = ???
}

trait Route
trait Printer
trait Marshaller[-A]  // must be contravariant

object Test {
  implicit def marshaller[A: Encoder](implicit p: Printer = ???): Marshaller[A] = ???
    // the `Printer` implicit arg seems to be necessary, either with default value, or no implicit in scope

  def foo[A](v: A)(implicit m: Marshaller[A]): Route = ???

  val route: Route = identity {
    val f: (Car => Route) => Route = ???  // ok if s/Car/Vehicle/
    f(vehicle => foo(vehicle))
  }
}
```

```
$ dotc -version
Scala compiler version 3.0.3-RC1-bin-SNAPSHOT-git-cc47c56 -- Copyright 2002-2021, LAMP/EPFL

$ dotc i13101.scala
exception occurred while typechecking i13101.scala
exception occurred while compiling i13101.scala
java.lang.AssertionError: assertion failed: Attempt to commit TS[203X, 18X, 16, 1, 0] into already committed TS[18X, 16, 1, 0] while compiling i13101.scala
Exception in thread "main" java.lang.AssertionError: assertion failed: Attempt to commit TS[203X, 18X, 16, 1, 0] into already committed TS[18X, 16, 1, 0]
	at scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:8)
	at dotty.tools.dotc.core.TyperState.commit(TyperState.scala:146)
	at dotty.tools.dotc.typer.Inferencing$.isFullyDefined(Inferencing.scala:40)
	at dotty.tools.dotc.typer.Implicits.ignoredInstanceNormalImport$1(Implicits.scala:881)
	at dotty.tools.dotc.typer.Implicits.$anonfun$1(Implicits.scala:896)
	at dotty.tools.dotc.typer.ImplicitSearchError.hiddenImplicitsAddendum(ErrorReporting.scala:419)
	at dotty.tools.dotc.typer.ImplicitSearchError.missingArgMsg(ErrorReporting.scala:267)
	at dotty.tools.dotc.typer.Implicits.missingArgMsg(Implicits.scala:897)
	at dotty.tools.dotc.typer.Implicits.missingArgMsg$(Implicits.scala:774)
	at dotty.tools.dotc.typer.Typer.missingArgMsg(Typer.scala:106)
	at dotty.tools.dotc.typer.Typer.issueErrors$1$$anonfun$1$$anonfun$1(Typer.scala:3353)
	at dotty.tools.dotc.reporting.NoExplanation.msg(Message.scala:132)
	at dotty.tools.dotc.reporting.Message.message(Message.scala:87)
	at dotty.tools.dotc.reporting.Message.isNonSensical(Message.scala:99)
	at dotty.tools.dotc.reporting.HideNonSensicalMessages.isHidden(HideNonSensicalMessages.scala:16)
	at dotty.tools.dotc.reporting.HideNonSensicalMessages.isHidden$(HideNonSensicalMessages.scala:10)
	at dotty.tools.dotc.reporting.AbstractReporter.isHidden(AbstractReporter.scala:8)
	at dotty.tools.dotc.reporting.Reporter.report(Reporter.scala:148)
	at dotty.tools.dotc.reporting.Reporter.flush$$anonfun$1(Reporter.scala:206)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at dotty.tools.dotc.reporting.Reporter.flush(Reporter.scala:206)
	at dotty.tools.dotc.core.TyperState.commit(TyperState.scala:143)
	at dotty.tools.dotc.typer.Applications.fail$1(Applications.scala:986)
	at dotty.tools.dotc.typer.Applications.realApply$7$$anonfun$6$$anonfun$2(Applications.scala:992)
	at scala.Option.getOrElse(Option.scala:201)
	at dotty.tools.dotc.typer.Applications.realApply$8$$anonfun$7(Applications.scala:993)
	at dotty.tools.dotc.typer.Typer.tryEither(Typer.scala:2952)
	at dotty.tools.dotc.typer.Applications.realApply$1(Applications.scala:994)
	at dotty.tools.dotc.typer.Applications.typedApply(Applications.scala:1032)
	at dotty.tools.dotc.typer.Applications.typedApply$(Applications.scala:317)
	at dotty.tools.dotc.typer.Typer.typedApply(Typer.scala:106)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2693)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2756)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2821)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2825)
	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:2941)
	at dotty.tools.dotc.typer.Typer.typedValDef(Typer.scala:2092)
	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2667)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2755)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2821)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2825)
	at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:2847)
	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:2897)
	at dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:2352)
	at dotty.tools.dotc.typer.Typer.typedTypeOrClassDef$2(Typer.scala:2681)
	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2685)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2755)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2821)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2825)
	at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:2847)
	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:2897)
	at dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:2475)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2726)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2756)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2821)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2825)
	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:2941)
	at dotty.tools.dotc.typer.FrontEnd.liftedTree1$1(FrontEnd.scala:79)
	at dotty.tools.dotc.typer.FrontEnd.typeCheck$$anonfun$1(FrontEnd.scala:84)
	at dotty.tools.dotc.typer.FrontEnd.monitor(FrontEnd.scala:43)
	at dotty.tools.dotc.typer.FrontEnd.typeCheck(FrontEnd.scala:85)
	at dotty.tools.dotc.typer.FrontEnd.runOn$$anonfun$3(FrontEnd.scala:120)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at dotty.tools.dotc.typer.FrontEnd.runOn(FrontEnd.scala:120)
	at dotty.tools.dotc.Run.runPhases$4$$anonfun$4(Run.scala:205)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
	at dotty.tools.dotc.Run.runPhases$5(Run.scala:216)
	at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:224)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:67)
	at dotty.tools.dotc.Run.compileUnits(Run.scala:231)
	at dotty.tools.dotc.Run.compileSources(Run.scala:166)
	at dotty.tools.dotc.Run.compile(Run.scala:150)
	at dotty.tools.dotc.Driver.doCompile(Driver.scala:39)
	at dotty.tools.dotc.Driver.process(Driver.scala:199)
	at dotty.tools.dotc.Driver.process(Driver.scala:167)
	at dotty.tools.dotc.Driver.process(Driver.scala:179)
	at dotty.tools.dotc.Driver.main(Driver.scala:209)
	at dotty.tools.dotc.Main.main(Main.scala)
```

--- a/compiler/src/dotty/tools/dotc/core/TyperState.scala
+++ b/compiler/src/dotty/tools/dotc/core/TyperState.scala
@@ -143,7 +143,15 @@ class TyperState() {
     reporter.flush()
     setCommittable(false)
     val targetState = ctx.typerState
-    assert(!targetState.isCommitted, s"Attempt to commit $this into already committed $targetState")
+
+    // Committing into an already committed TyperState usually doesn't make
+    // sense since it means the constraints we're committing won't be propagated
+    // further, but it can happen if the targetState gets captured in a reported
+    // Message, because forcing that Message might involve creating and
+    // committing new TyperStates into the captured one after its been committed.
+    assert(!targetState.isCommitted || targetState.reporter.hasErrors || targetState.reporter.hasWarnings,
+      s"Attempt to commit $this into already committed $targetState")
+
     if constraint ne targetState.constraint then
       Stats.record("typerState.commit.new constraint")
       constr.println(i"committing $this to $targetState, fromConstr = $constraint, toConstr = ${targetState.constraint}")

--- a/tests/neg/i13101.scala
+++ b/tests/neg/i13101.scala
@@ -1,0 +1,24 @@
+trait Vehicle
+trait Car extends Vehicle
+
+trait Encoder[A]
+object Encoder {
+  implicit val encodeVehicle: Encoder[Vehicle] = ???
+  implicit val encodeCar: Encoder[Car] = ???
+}
+
+trait Route
+trait Printer
+trait Marshaller[-A]  // must be contravariant
+
+object Test {
+  implicit def marshaller[A: Encoder](implicit p: Printer = ???): Marshaller[A] = ???
+    // the `Printer` implicit arg seems to be necessary, either with default value, or no implicit in scope
+
+  def foo[A](v: A)(implicit m: Marshaller[A]): Route = ???
+
+  val route: Route = identity {
+    val f: (Car => Route) => Route = ???  // ok if s/Car/Vehicle/
+    f(vehicle => foo(vehicle)) // error: ambiguous implicit
+  }
+}

